// verify-aura-backend/src/services/statsService.ts
//backend/src/services/statsService.ts

import { supabase, cachedQuery, supabasePool } from "./supabaseOptimized";

export const getAdminStats = async () => {
  try {
    // Use cached queries for frequently accessed stats
    const statsData = await cachedQuery(
      'admin-stats',
      async (client) => {
        // Query all stats in parallel using the optimized pool
        const [
          eventsQuery,
          participantsQuery,
          logsQuery,
          revokedQuery,
          recentLogsQuery
        ] = await Promise.all([
          client.from("events").select("*", { count: "exact" }),
          client.from("participants").select("*", { count: "exact" }),
          client.from("activity_logs").select("*", { count: "exact" }),
          client.from("participants").select("id").eq("revoked", true),
          client
            .from("activity_logs")
            .select("id, action, user_email, created_at, metadata")
            .order("created_at", { ascending: false })
            .limit(5)
        ]);
        
        return {
          eventsQuery,
          participantsQuery,
          logsQuery,
          revokedQuery,
          recentLogsQuery
        };
      },
      30000 // Cache for 30 seconds
    );
    
    const { eventsQuery, participantsQuery, logsQuery, revokedQuery, recentLogsQuery } = statsData;

    // Handle errors
    if (eventsQuery.error) throw eventsQuery.error;
    if (participantsQuery.error) throw participantsQuery.error;
    if (logsQuery.error) throw logsQuery.error;
    if (revokedQuery.error) throw revokedQuery.error;
    if (recentLogsQuery.error) throw recentLogsQuery.error;

    const totalEvents = eventsQuery.count || 0;
    const totalParticipants = participantsQuery.count || 0;
    const totalRevoked = revokedQuery?.data?.length || 0;
    const totalActiveCertificates = totalParticipants - totalRevoked;
    const recentActivitiesCount = logsQuery.count || 0;

    return {
      total_events: totalEvents,
      active_events: totalEvents, // For now, consider all events as active
      total_participants: totalParticipants,
      total_revoked: totalRevoked,
      total_active_certificates: totalActiveCertificates,
      recent_activities_count: recentActivitiesCount,
      recent_activities: recentLogsQuery.data || [],
    };
  } catch (error) {
    console.error('[STATS SERVICE ERROR]', error);
    // Clear cache on error to force fresh data on next request
    throw new Error('Failed to load statistics');
  }
};

// // Alternative implementation (commented out)
// // Uses optimized imports if needed later
// import { supabase } from "./supabaseOptimized";

// // Utility to calculate date 7 days ago in ISO format
// const get7DaysAgoISO = () => {
//   const now = new Date();
//   now.setDate(now.getDate() - 7);
//   return now.toISOString();
// };

// export const getAdminStats = async () => {
//   const sevenDaysAgo = get7DaysAgoISO();

//   // Query multiple stats in parallel
//   const [
//     { count: totalEvents },
//     { data: upcomingEvents },
//     { count: totalParticipants },
//     { data: revokedParticipants },
//     { data: recentLogs },
//     { data: logsLast7Days },
//   ] = await Promise.all([
//     supabase.from("events").select("*", { count: "exact", head: true }),

//     supabase
//       .from("events")
//       .select("id, date")
//       .gte("date", new Date().toISOString().split("T")[0]),

//     supabase.from("participants").select("*", { count: "exact", head: true }),

//     supabase.from("participants").select("id").eq("revoked", true),

//     supabase
//       .from("activity_logs")
//       .select("id, action, user_email, created_at, metadata")
//       .order("created_at", { ascending: false })
//       .limit(5),

//     supabase
//       .from("activity_logs")
//       .select("id")
//       .gte("created_at", sevenDaysAgo),
//   ]);

//   const revokedCount = revokedParticipants?.length || 0;
//   const activeCertificates = (totalParticipants || 0) - revokedCount;
//   const recentActivitiesCount = logsLast7Days?.length || 0;

//   return {
//     total_events: totalEvents || 0,
//     active_events: upcomingEvents?.length || 0,
//     total_participants: totalParticipants || 0,
//     total_revoked: revokedCount,
//     total_active_certificates: activeCertificates,
//     recent_activities_count: recentActivitiesCount,
//     recent_activities: recentLogs || [],
//   };
// };
