// verify-aura-backend/src/routes/admin.ts
//backend/src/routes/admin.ts (for event routes)

import express from 'express';
import { supabase, supabasePool, cachedQuery } from '../services/supabaseOptimized';
import { requireAdminOptimized as requireAdmin } from '../middleware/clerkAuthOptimized';
import { QueryBuilderOptimized } from '../utils/queryBuilderOptimized';
import { QueryParser } from '../utils/queryParser';
import { importParticipantsFromGoogleSheets, getImportStatus } from '../services/googleSheetsImport';

const router = express.Router();
const queryBuilder = new QueryBuilderOptimized(supabase);

/**
 * Root Admin Info
 */
router.get('/', requireAdmin, (_req, res) => {
  res.json({
    message: 'Admin API root',
    routes: {
      events: '/api/admin/events',
      participants: '/api/admin/participants',
      logs: '/api/admin/logs',
      health: '/api/admin/health',
    },
  });
});

/**
 * Health Check
 */
router.get('/health', requireAdmin, async (_req, res) => {
  try {
    const { error } = await supabase.from('events').select('id').limit(1);
    if (error) throw error;

    return res.json({
      status: 'ok',
      db: 'reachable',
      timestamp: new Date().toISOString(),
    });
  } catch (err: any) {
    return res.status(500).json({
      status: 'error',
      db: 'unreachable',
      error: err.message || String(err),
      timestamp: new Date().toISOString(),
    });
  }
});

/**
 * GET /api/admin/events
 * Advanced filtering, sorting, and pagination for events
 * 
 * Query Parameters:
 * - search: string (searches event_name, event_code, description)
 * - tag: string (filter by event tag)
 * - created_by: string (filter by creator)
 * - date_from: string (ISO date)
 * - date_to: string (ISO date)
 * - event_status: 'upcoming' | 'ongoing' | 'ended'
 * - sort_by: string (field to sort by)
 * - sort_order: 'asc' | 'desc'
 * - page: number
 * - limit: number
 */
router.get('/events', requireAdmin, async (req, res) => {
  try {
    // Parse and validate query parameters
    const { filters, sort, pagination, errors } = QueryParser.parseQueryParams(req);

    if (errors.length > 0) {
      console.log('[EVENTS QUERY] Validation errors:', errors);
      return res.status(400).json({
        success: false,
        message: 'Invalid query parameters',
        errors,
        allowedSortFields: QueryParser.getAllowedSortFields('events')
      });
    }

    console.log('[EVENTS QUERY] Executing with:', { filters, sort, pagination });

    // Execute query
    const result = await queryBuilder.queryEvents(filters, sort, pagination);

    console.log('[EVENTS QUERY] Success, returned', result.data?.length, 'events');

    return res.status(200).json(result);

  } catch (error) {
    console.error('[EVENTS QUERY ERROR]', error);
    console.error('[EVENTS QUERY ERROR] Stack:', error instanceof Error ? error.stack : 'No stack trace');
    return res.status(500).json({
      success: false,
      message: 'Failed to fetch events',
      code: 'EVENTS_FETCH_FAILED',
      error: error instanceof Error ? error.message : 'Database error'
    });
  }
});

// router.get('/events', requireAdmin, async (req, res) => {
//   try {
//     // Parse and validate query parameters
//     const { filters, sort, pagination, errors } = QueryParser.parseQueryParams(req);

//     // Return validation errors if any
//     if (errors.length > 0) {
//       return res.status(400).json({
//         success: false,
//         message: 'Invalid query parameters',
//         errors: errors,
//         allowedSortFields: QueryParser.getAllowedSortFields('events')
//       });
//     }

//     // Sanitize search term if provided
//     if (filters.search) {
//       filters.search = QueryParser.sanitizeSearchTerm(filters.search);
//     }

//     // Execute query using QueryBuilder
//     const result = await queryBuilder.queryEvents(filters, sort, pagination);

//     // Log the query for debugging
//     console.log(`[EVENTS QUERY] ${QueryParser.buildQueryString({ filters, sort, pagination })} - ${result.meta.queryTime}ms`);

//     return res.status(200).json({
//       success: true,
//       ...result
//     });

//   } catch (error) {
//     console.error('[EVENTS QUERY ERROR]', error);
//     return res.status(500).json({
//       success: false,
//       message: 'Failed to fetch events',
//       error: error instanceof Error ? error.message : 'Unknown error'
//     });
//   }
// });

/**
 * GET /api/admin/events/:id
 * Get single event by ID
 */
router.get('/events/:id', requireAdmin, async (req, res) => {
  try {
    const { id } = req.params;

    // Validate UUID format
    if (!QueryParser['isValidUUID'](id)) {
      return res.status(400).json({
        success: false,
        message: 'Invalid event ID format'
      });
    }

    const { data, error } = await supabase
      .from('events')
      .select('*')
      .eq('id', id)
      .single();

    if (error || !data) {
      return res.status(404).json({
        success: false,
        message: 'Event not found'
      });
    }

    return res.status(200).json(data);

  } catch (error) {
    console.error('[GET EVENT ERROR]', error);
    return res.status(500).json({
      success: false,
      message: 'Failed to fetch event',
      error: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * POST /api/admin/events
 * Create new event
 */
router.post('/events', requireAdmin, async (req, res) => {
  try {
    const {
      event_name,
      event_code,
      date,
      google_sheet_url,
      tag,
      description,
    } = req.body;

    // Validate required fields
    if (!event_name || !event_code || !date) {
      return res.status(400).json({
        success: false,
        message: 'event_name, event_code, and date are required'
      });
    }

    // Validate date format
    const eventDate = new Date(date);
    if (isNaN(eventDate.getTime())) {
      return res.status(400).json({
        success: false,
        message: 'Invalid date format'
      });
    }

    const created_by = (req as any).auth?.userId || 'system';

    const { data, error } = await supabase
      .from('events')
      .insert([{
        event_name,
        event_code,
        date,
        google_sheet_url,
        tag,
        description,
        created_by,
      }])
      .select()
      .single();

    if (error) {
      return res.status(400).json({
        success: false,
        message: 'Failed to create event',
        error: error.message
      });
    }

    // Log the action
    await supabase.from('activity_logs').insert([{
      user_id: created_by,
      user_email: (req as any).auth?.email || null,
      action: 'event_created',
      metadata: {
        event_name: data.event_name,
        event_code: data.event_code,
      },
    }]);

    return res.status(201).json(data);

  } catch (error) {
    console.error('[CREATE EVENT ERROR]', error);
    return res.status(500).json({
      success: false,
      message: 'Failed to create event',
      error: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * PUT /api/admin/events/:id
 * Update event
 */
router.put('/events/:id', requireAdmin, async (req, res) => {
  try {
    const { id } = req.params;

    // Validate UUID format
    if (!QueryParser['isValidUUID'](id)) {
      return res.status(400).json({
        success: false,
        message: 'Invalid event ID format'
      });
    }

    const updated_by = (req as any).auth?.userId || 'system';

    // Check if event exists
    const { data: existingEvent, error: checkError } = await supabase
      .from('events')
      .select('id, event_name, event_code')
      .eq('id', id)
      .single();

    if (checkError || !existingEvent) {
      return res.status(404).json({
        success: false,
        message: 'Event not found'
      });
    }

    const updates = {
      ...req.body,
      updated_by,
    };

    const { data, error } = await supabase
      .from('events')
      .update(updates)
      .eq('id', id)
      .select()
      .single();

    if (error) {
      return res.status(400).json({
        success: false,
        message: 'Failed to update event',
        error: error.message
      });
    }

    // Log the action
    await supabase.from('activity_logs').insert([{
      user_id: updated_by,
      user_email: (req as any).auth?.email || 'system',
      action: 'event_updated',
      metadata: {
        event_name: data.event_name,
        event_code: data.event_code,
        updates: req.body,
      },
    }]);

    return res.status(200).json(data);

  } catch (error) {
    console.error('[UPDATE EVENT ERROR]', error);
    return res.status(500).json({
      success: false,
      message: 'Failed to update event',
      error: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * DELETE /api/admin/events/:id
 * Delete event
 */
router.delete('/events/:id', requireAdmin, async (req, res) => {
  try {
    const { id } = req.params;

    // Validate UUID format
    if (!QueryParser['isValidUUID'](id)) {
      return res.status(400).json({
        success: false,
        message: 'Invalid event ID format'
      });
    }

    // Fetch event details before deletion
    const { data: event, error: fetchError } = await supabase
      .from('events')
      .select('event_name, event_code')
      .eq('id', id)
      .single();

    if (fetchError || !event) {
      return res.status(404).json({
        success: false,
        message: 'Event not found'
      });
    }

    // Check if event has participants
    const { data: participants, error: participantsError } = await supabase
      .from('participants')
      .select('id')
      .eq('event_id', id)
      .limit(1);

    if (participantsError) {
      console.error('Error checking participants:', participantsError);
      return res.status(500).json({
        success: false,
        message: 'Error checking event participants'
      });
    }

    if (participants && participants.length > 0) {
      return res.status(400).json({
        success: false,
        message: 'Cannot delete event with participants. Please remove all participants first.'
      });
    }

    const { error: deleteError } = await supabase
      .from('events')
      .delete()
      .eq('id', id);

    if (deleteError) {
      return res.status(400).json({
        success: false,
        message: 'Failed to delete event',
        error: deleteError.message
      });
    }

    // Log the action
    await supabase.from('activity_logs').insert([{
      user_id: (req as any).auth?.userId || 'system',
      user_email: (req as any).auth?.email || null,
      action: 'event_deleted',
      metadata: {
        event_name: event.event_name,
        event_code: event.event_code,
      },
    }]);

    return res.status(204).send();

  } catch (error) {
    console.error('[DELETE EVENT ERROR]', error);
    return res.status(500).json({
      success: false,
      message: 'Failed to delete event',
      error: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * POST /api/admin/events/:id/sync-participants
 * Sync participants from Google Sheets for a specific event
 */
router.post('/events/:id/sync-participants', requireAdmin, async (req, res) => {
  try {
    const { id: eventId } = req.params;
    const { google_sheet_url } = req.body;

    // Validate UUID format
    if (!QueryParser['isValidUUID'](eventId)) {
      return res.status(400).json({
        success: false,
        message: 'Invalid event ID format'
      });
    }

    // Validate Google Sheet URL
    if (!google_sheet_url || typeof google_sheet_url !== 'string') {
      return res.status(400).json({
        success: false,
        message: 'Google Sheet URL is required'
      });
    }

    // Check if URL is a valid Google Apps Script URL
    if (!google_sheet_url.includes('script.google.com/macros/s/')) {
      return res.status(400).json({
        success: false,
        message: 'Invalid Google Apps Script URL. Please provide a valid Google Apps Script web app URL.'
      });
    }

    // Check if event exists
    const { data: event, error: eventError } = await supabase
      .from('events')
      .select('id, event_name, sync_status')
      .eq('id', eventId)
      .single();

    if (eventError || !event) {
      return res.status(404).json({
        success: false,
        message: 'Event not found'
      });
    }

    // Check if sync is already in progress
    if (event.sync_status === 'syncing') {
      return res.status(409).json({
        success: false,
        message: 'Sync is already in progress for this event. Please wait for it to complete.'
      });
    }

    const userId = (req as any).auth?.userId || 'system';
    const userEmail = (req as any).auth?.email || null;

    console.log(`[SYNC] Starting participant sync for event ${eventId}`);
    
    // Perform the import
    const result = await importParticipantsFromGoogleSheets(
      eventId,
      google_sheet_url,
      userId,
      userEmail
    );

    console.log(`[SYNC] Completed sync for event ${eventId}:`, {
      imported: result.imported,
      updated: result.updated,
      skipped: result.skipped,
      errors: result.errors
    });

    return res.status(200).json({
      success: result.success,
      message: `Successfully synced participants from Google Sheets`,
      data: result
    });

  } catch (error) {
    console.error('[SYNC ERROR]', error);
    
    // Determine appropriate error status
    const statusCode = error instanceof Error && error.message === 'Event not found' ? 404 : 500;
    
    return res.status(statusCode).json({
      success: false,
      message: 'Failed to sync participants',
      error: error instanceof Error ? error.message : 'Unknown error',
      code: 'SYNC_FAILED'
    });
  }
});

/**
 * GET /api/admin/events/:id/sync-status
 * Get the current sync status of an event
 */
router.get('/events/:id/sync-status', requireAdmin, async (req, res) => {
  try {
    const { id: eventId } = req.params;

    // Validate UUID format
    if (!QueryParser['isValidUUID'](eventId)) {
      return res.status(400).json({
        success: false,
        message: 'Invalid event ID format'
      });
    }

    const status = await getImportStatus(eventId);

    return res.status(200).json({
      success: true,
      data: status
    });

  } catch (error) {
    console.error('[SYNC STATUS ERROR]', error);
    return res.status(500).json({
      success: false,
      message: 'Failed to get sync status',
      error: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * GET /api/admin/events/:id/participants-count
 * Get the count of participants for an event
 */
router.get('/events/:id/participants-count', requireAdmin, async (req, res) => {
  try {
    const { id: eventId } = req.params;

    // Validate UUID format
    if (!QueryParser['isValidUUID'](eventId)) {
      return res.status(400).json({
        success: false,
        message: 'Invalid event ID format'
      });
    }

    const { count, error } = await supabase
      .from('participants')
      .select('*', { count: 'exact', head: true })
      .eq('event_id', eventId);

    if (error) {
      throw error;
    }

    return res.status(200).json({
      success: true,
      data: {
        total: count || 0
      }
    });

  } catch (error) {
    console.error('[PARTICIPANTS COUNT ERROR]', error);
    return res.status(500).json({
      success: false,
      message: 'Failed to get participants count',
      error: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

export default router;
